File: Modul_3/Lesson_3.1(DFS, BFS)_HW/1_cell_removal(1317).py
def is_board(v, u):
    return 0 <= v < n and 0 <= u < m

def dfs(v, u, used):
    used[v][u] = True
    for i in range(4):
        new_u = u + dx[i]
        new_v = v + dy[i]
        if (is_board(new_v, new_u)):
            if graph[new_v][new_u] == 1 and used[new_v][new_u] == False:
                dfs(new_v, new_u, used)



n, m = map(int, input().split())
used = [[False] * m for _ in range(n)]
graph = [[] * m for _ in range(n)]
for i in range(n):
    a = list(input())
    for j in range(m):
        if a[j] == "#":
            graph[i].append(1)
        else:
            graph[i].append(0)


dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

cnt = 0
for i in range(n):
    for j in range(m):
        if graph[i][j] == 1 and used[i][j] == False:
            dfs(i, j, used)
            cnt += 1


print(cnt)

File: Modul_3/Lesson_3.1(DFS, BFS)_HW/1_cell_removal(1317).cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

int n, m;
vector <vector <bool>> used;
vector <vector <int>> graph;
int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, -1, 1};

void read() {
    cin >> n >> m;
    used.resize(n, vector <bool> (m, false));
    graph.resize(n, vector <int> (m, 0));
    for (int i = 0; i < n; i++) {
        string line;
        cin >> line;
        for (int j = 0; j < m; j++) {
            if (line[j] == '#') graph[i][j] = 1;
            
        }
    }
}
bool is_board(int i, int j){
    return (0 <= i && i < n && 0 <= j && j < m);
}
void dfs(int i, int j, vector < vector <bool>> &used){
    used[i][j] = true;
    for (int k = 0; k < 4; k++) {
        int new_i = i + dy[k];
        int new_j = j + dx[k];
        if (is_board(new_i, new_j)){
            if (graph[new_i][new_j] == 1 && used[new_i][new_j] == false){
                dfs(new_i, new_j, used);
            }
        }
    }

}
void solve() {
    int ans = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++){
            if (graph[i][j] == 1 && used[i][j] == false) {
                dfs(i, j, used);
                ans++;
            }
            
        }
    }
    
    cout << ans << endl;
}




int main(){
    read();
    solve();
    return 0;
}





